src/coreclr/jit/layout.cpp

```diff
@@ -419,6 +426,11 @@ bool ClassLayout::AreCompatible(const ClassLayout* layout1, const ClassLayout* l
         return false;
     }
 
+    if (layout1->GetType() != layout2->GetType())
+    {
+        return false;
+    }
```

Comment:
Can we assert this in the next `then` instead? It's implied by the sizes being the same if they do not have GC pointers, right?

Reply:
And also when they do have GC pointers I assume (both should be `TYP_STRUCT`). In fact, it seems kind of redundant? Can't we just get this from `HasGCPtr()` and `GetSize()` by calling `getSIMDTypeForSize` which should not be that expensive?

Reply:
Do you mean like this?
```cs
    if (!layout1->HasGCPtr() && !layout2->HasGCPtr())
    {
        return true;
    }
    else
    {
         assert(layout1->GetType() == layout2->GetType());
    }
```
?

The intent of this check is to disallow compatibility between `struct<16>` and `simd16` (which what the assert in #69232 was complaining about).

Reply:
> The intent of this check is to disallow compatibility between struct<16> and simd16

It is confusing to me that such two layouts are incompatible when neither have GC pointers. It is my understanding that `ClassLayout::AreCompatible` has primarily been used for copy purposes before, are we starting to use it for ABI purposes as well?

Reply:
> It is confusing to me that such two layouts are incompatible when neither have GC pointers.

Indeed, it is a questionable behavior. One can "justify" it by asking a question akin to "are `TYP_INT` variables "compatible" with `struct<4>` variables"?

However, this behavior is also pre-existing. Today, if you do this:
```cs
simd16 b = { ... };
struct<16> a = *(struct<16>*)&b;
```
You will get this tree after block morphing:
```scala
Local V01 should not be enregistered because: cast takes addr
MorphCopyBlock (after):
               [000009] -A-X-------                         *  ASG       struct (copy)
               [000008] ---X-+-N---                         +--*  BLK       struct<RyuJitReproduction.Program+Struct16, 16>
               [000007] -----+-----                         |  \--*  LCL_VAR   byref  V00 RetBuf       
               [000006] n----+-----                         \--*  OBJ       struct<RyuJitReproduction.Program+Struct16, 16>
               [000005] -----+-----                            \--*  ADDR      long  
               [000004] -----+-N---                               \--*  LCL_VAR   simd16<System.Runtime.Intrinsics.Vector128`1[System.Int32]> V01 loc0         
```
(You can note that `cast takes addr` DNER reason)

So, this purely reflects what behavior the of block copying is today.

We should of course fix this pessimization; but it's not yet totally clear how (canonicalizing to `TYP_STRUCT` / to `TYP_SIMD` / just fully supporting the mismatch -- but then we should perhaps support mismatches with all primitives...?)

Reply:
Alright, in that case this looks fine to me.